File: wpimath/src/test/java/edu/wpi/first/math/kinematics/MecanumDriveMotorVoltagesSerdeTest.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package edu.wpi.first.math.kinematics;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertEquals;

import edu.wpi.first.math.proto..ProtobufMecanumDriveMotorVoltages;
import edu.wpi.first.util.struct.Struct;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import org.junit.jupiter.api.Test;
import java.util.Arrays;

public class MecanumDriveMotorVoltagesSerdeTest {
  private static final MecanumDriveMotorVoltages DATA = new MecanumDriveMotorVoltages(1.91, 2.29);
  private static final byte[] STRUCT_BUFFER = new byte[]{63, -2, -113, 92, 40, -11, -62, -113, 64, 2, 81, -21, -123, 30, -72, 82};

  @Test
  void testStructPack() {
    ByteBuffer buffer = ByteBuffer.allocate(MecanumDriveMotorVoltages.struct.getSize());
    buffer.order(ByteOrder.LITTLE_ENDIAN);
    MecanumDriveMotorVoltages.struct.pack(buffer, DATA);

    byte[] actual = buffer.array();
    String newContent = new String(buffer.array());
    System.out.println(Arrays.toString(actual));
    System.out.println(newContent);
    assertArrayEquals(actual, STRUCT_BUFFER);
  }

  @Test
  void testStructUnpack() {
    ByteBuffer buffer = ByteBuffer.wrap(STRUCT_BUFFER);
    buffer.order(ByteOrder.LITTLE_ENDIAN);

    MecanumDriveMotorVoltages data = MecanumDriveMotorVoltages.struct.unpack(buffer);
    assertEquals(DATA.frontLeftVoltage, data.frontLeftVoltage);
    assertEquals(DATA.frontRightVoltage, data.frontRightVoltage);
    assertEquals(DATA.rearLeftVoltage, data.rearLeftVoltage);
    assertEquals(DATA.rearRightVoltage, data.rearRightVoltage);
  }

  @Test
  void testProtoPack() {
    ProtobufMecanumDriveMotorVoltages proto = MecanumDriveMotorVoltages.proto.createMessage();
    MecanumDriveMotorVoltages.proto.pack(proto, DATA);

    assertEquals(DATA.frontLeftVoltage, proto.getFrontLeftVolts());
    assertEquals(DATA.frontRightVoltage, proto.getFrontRightVolts());
    assertEquals(DATA.rearLeftVoltage, proto.getRearLeftVolts());
    assertEquals(DATA.rearRightVoltage, proto.getRearRightVolts());
  }

  @Test
  void testProtoUnpack() {
    ProtobufMecanumDriveMotorVoltages proto = MecanumDriveMotorVoltages.proto.createMessage();
    proto.setFrontLeftVolts(DATA.frontLeftVoltage);
    proto.setFrontRightVolts(DATA.frontRightVoltage);
    proto.setRearLeftVolts(DATA.rearLeftVoltage);
    proto.setRearRightVolts(DATA.rearRightVoltage);

    MecanumDriveMotorVoltages data = MecanumDriveMotorVoltages.proto.unpack(proto);
    assertEquals(DATA.frontLeftVoltage, data.frontLeftVoltage);
    assertEquals(DATA.frontRightVoltage, data.frontRightVoltage);
    assertEquals(DATA.rearLeftVoltage, data.rearLeftVoltage);
    assertEquals(DATA.rearRightVoltage, data.rearRightVoltage);
  }
}

File: wpimath/src/test/native/cpp/kinematics/serde/MecanumDriveMotorVoltagesSerdeTest.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include <gtest/gtest.h>

#include "frc/geometry/MecanumDriveMotorVoltages.h"
#include "geometry2d.pb.h"

using namespace frc;

namespace {

using StructType = wpi::Struct<frc::MecanumDriveMotorVoltages>;
using ProtoType = wpi::Protobuf<frc::MecanumDriveMotorVoltages>;

constexpr std::array<uint8_t, StructType::kSize> create_test_buffer() {
  std::array<uint8_t, StructType::kSize> output;
  int buffer[] = {
    0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0,
};
  for (size_t idx = 0; idx < StructType::kSize; ++idx) {
    output[idx] = static_cast<uint8_t>(buffer[idx]);
  }
  return output;
}

std::array<uint8_t, StructType::kSize> kExpectedStructBuffer =
    create_test_buffer();

constexpr MecanumDriveMotorVoltages kExpectedData{};
}  // namespace

TEST(MecanumDriveMotorVoltagesTest, StructPack) {
  uint8_t buffer[StructType::kSize];
  StructType::Pack(buffer, kExpectedData);

  for (size_t i = 0; i < StructType::kSize; ++i) {
    EXPECT_EQ(kExpectedStructBuffer[i], buffer[i]) << " on byte " << i;
  }
}

TEST(MecanumDriveMotorVoltagesTest, StructUnpack) {
  MecanumDriveMotorVoltages unpacked_data = StructType::Unpack(kExpectedStructBuffer);

    EXPECT_EQ(kExpectedData.FrontLeftVolts(), unpacked_data.FrontLeftVolts());
    EXPECT_EQ(kExpectedData.FrontRightVolts(), unpacked_data.FrontRightVolts());
    EXPECT_EQ(kExpectedData.RearLeftVolts(), unpacked_data.RearLeftVolts());
    EXPECT_EQ(kExpectedData.RearRightVolts(), unpacked_data.RearRightVolts());
}


TEST(MecanumDriveMotorVoltagesTest, ProtobufPack) {
  wpi::proto::ProtobufMecanumDriveMotorVoltages proto;
  ProtoType::Pack(&proto, kExpectedData);

    EXPECT_EQ(kExpectedData.FrontLeftVolts(), proto.front_left_volts());
    EXPECT_EQ(kExpectedData.FrontRightVolts(), proto.front_right_volts());
    EXPECT_EQ(kExpectedData.RearLeftVolts(), proto.rear_left_volts());
    EXPECT_EQ(kExpectedData.RearRightVolts(), proto.rear_right_volts());
}



TEST(MecanumDriveMotorVoltagesTest, ProtobufUnpack) {
  wpi::proto::ProtobufMecanumDriveMotorVoltages proto;
  proto.set_front_left_volts(kExpectedData.FrontLeftVolts());
  proto.set_front_right_volts(kExpectedData.FrontRightVolts());
  proto.set_rear_left_volts(kExpectedData.RearLeftVolts());
  proto.set_rear_right_volts(kExpectedData.RearRightVolts());

  MecanumDriveMotorVoltages unpacked_data = ProtoType::Unpack(proto);
    EXPECT_EQ(kExpectedData.FrontLeftVolts(), unpacked_data.FrontLeftVolts());
    EXPECT_EQ(kExpectedData.FrontRightVolts(), unpacked_data.FrontRightVolts());
    EXPECT_EQ(kExpectedData.RearLeftVolts(), unpacked_data.RearLeftVolts());
    EXPECT_EQ(kExpectedData.RearRightVolts(), unpacked_data.RearRightVolts());
}